File: 2 add two numbers/javascript/add-two.js
'use strict';

function ListNode(val) {
  this.val = val;
  this.next = null;
}

/**
 * @param {ListNode} l1
 * @param {ListNode} l2
 * @return {ListNode}
 */
var addTwoNumbers = function (l1, l2) {
  let list = null,
    head = null
  let remain = 0

  while (l1 !== null || l2 !== null) {
    let sum = 0
    if (l1 !== null) {
      sum += l1.val
      l1 = l1.next
    }

    if (l2 != null) {
      sum += l2.val
      l2 = l2.next
    }

    sum += remain
    if (sum >= 10) {
      remain = Math.floor(sum / 10)
      sum -= remain * 10
    } else {
      remain = 0
    }

    var newVal = new ListNode(sum)
    if (list === null) {
      list = newVal
      head = newVal
    } else {
      head.next = newVal
      head = head.next
    }
  }

  if (remain !== 0) {
    var newVal = new ListNode(remain)
    head.next = newVal
    head = head.next
  }

  return list
};

File: 2 add two numbers/c/add-two.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */

struct ListNode {
  int val;
  ListNode *next;
  ListNode(int x) : val(x), next(NULL) {}
};

int advance(ListNode *&n)
{
	if (n == nullptr) return 0;

	int v = n->val;
	n = n->next;
	return v;
}

class Solution
{
public:
	ListNode *addTwoNumbers(ListNode *l1, ListNode *l2)
	{
		auto list = new ListNode(0), head = list;
		auto p = l1, q = l2;
		int carry = 0;

		while (p != nullptr || q != nullptr)
		{
			int sum = advance(p) + advance(q) + carry;
			carry = sum / 10;
			sum -= carry * 10;
			head->next = new ListNode(sum);
			head = head->next;
		}

		if (carry > 0)
		{
			head->next = new ListNode(carry);
		}

		return list->next;
	}
};
